### DISCUSSION TITLE
Updating record

### DISCUSSION THREAD

**Discussion by fctoscano** [phalcon/phalcon#349] [Q&A]
Hi.

Lemme try to explain what´s happening.

I have a table with the fields below:

``service_orders` (
  `id` int unsigned NOT NULL AUTO_INCREMENT,
  `service_order_number` int unsigned NOT NULL DEFAULT '0',
  `service_order_type_id` int unsigned NOT NULL DEFAULT '0',
  `equipment_id` int unsigned NOT NULL DEFAULT '0',
  `address` varchar(328) COLLATE utf8mb4_general_ci NOT NULL DEFAULT '0',
  `contact` varchar(60) COLLATE utf8mb4_general_ci DEFAULT NULL,
  `distance` int unsigned DEFAULT NULL,
  `time` int unsigned DEFAULT NULL,
  `occurence` mediumtext COLLATE utf8mb4_general_ci,
  `technical_report` mediumtext COLLATE utf8mb4_general_ci,
  `client_observation` mediumtext COLLATE utf8mb4_general_ci,
  `technician` varchar(60) COLLATE utf8mb4_general_ci DEFAULT NULL,
  `client_name` varchar(60) COLLATE utf8mb4_general_ci DEFAULT NULL,
  `hours_date` date DEFAULT NULL,
  `hours` int unsigned DEFAULT NULL,
  `kilometers_date` date DEFAULT NULL,
  `kilometers` int unsigned DEFAULT NULL,
  `closure_date` timestamp NULL DEFAULT NULL,
  `status` enum('A','F','C','E') COLLATE utf8mb4_general_ci NOT NULL DEFAULT 'A',
  `user_id` int unsigned DEFAULT NULL,
  `created_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP,
  `updated_at` timestamp NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (`id`),
  KEY `FK_service_orders_service_order_types` (`service_order_type_id`),
  KEY `FK_service_orders_equipments` (`equipment_id`),
  KEY `FK_service_orders_users` (`user_id`),
  CONSTRAINT `FK_service_orders_equipments` FOREIGN KEY (`equipment_id`) REFERENCES `equipments` (`id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `FK_service_orders_service_order_types` FOREIGN KEY (`service_order_type_id`) REFERENCES `service_order_types` (`id`) ON DELETE CASCADE ON UPDATE CASCADE,
  CONSTRAINT `FK_service_orders_users` FOREIGN KEY (`user_id`) REFERENCES `users` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2039 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci;`

Here I have the model

`<?php
declare(strict_types=1);

namespace Smartman\Models;

use Phalcon\Mvc\Model;
//use Phalcon\Filter\Validation;
//use Phalcon\Filter\Validation\Validator\Uniqueness;

/**
* Smartman\Models\ServiceOrders
* ServiceOrders per account
*/
class ServiceOrders extends Model
{

    /**
    * @var integer
    */
    public $id;

    /**
    * @var integer
    */
    public $service_order_number;

    /**
    * @var integer
    */
    public $service_order_type_id;

    /**
    * @var integer
    */
    public $equipment_id;

    /**
    * @var string
    */
    public $address;

    /**
    * @var string
    */
    public $contact;

    /**
    * @var float
    */
    public $distance;

    /**
    * @var string
    */
    public $time;

    /**
    * @var string
    */
    public $occurence;

    /**
    * @var string
    */
    public $technical_report;

    /**
    * @var string
    */
    public $client_observation;

    /**
    * @var string
    */
    public $technician;

    /**
    * @var string
    */
    public $client_name;

    /**
    * @var string
    */
    public $hours_date;

    /**
    * @var string
    */
    public $hours;

    /**
    * @var string
    */
    public $kilometers_date;

    /**
    * @var string
    */
    public $kilometers;

    /**
    * @var string
    */
    public $closure_date;

    /**
    * @var string
    */
    public $status;

    /**
    * @var integer
    */
    public $user_id;

    /**
    * @var string
    */
    public $created_at;

    /**
    * @var string
    */
    public $updated_at;

    public function initialize()
    {
        $this->belongsTo(
            'service_order_type_id',
            'Smartman\Models\ServiceOrderTypes',
            'id',
            array('alias' => 'serviceordertype')
        );

        $this->belongsTo(
            'equipment_id',
            'Smartman\Models\Equipments',
            'id',
            array('alias' => 'equipment')
        );

        $this->hasMany(
            'id',
            'Smartman\Models\ServiceOrderDates',
            'service_order_id',
            array('alias' => 'serviceorderdates')
        );

        // Skips fields/columns on both INSERT/UPDATE operations
        $this->skipAttributes(
            [
                "created_at",
                "updated_at",
            ]
        );
    }
}`

I have two methods in the controller, create and edit.

Whaen I create a new record (service order) it  goes well, but when I edit this record, it updates all the fields except the service_order_type_id.

Here´s the edit method:

`public function editAction($id)
    {
        $serviceorder = ServiceOrders::findFirstById($id);

        if ($this->request->isPost()) {

            // BUG: not saving service_order_type_id
            $serviceorder->assign([
                'service_order_type_id' => $this->request->getPost('serviceordertype'),
                'equipment_id'          => $this->request->getPost('equipment'),
                'occurence'             => $this->request->getPost('occurence'),
                'technical_report'      => $this->request->getPost('technicalreport'),
                'client_observation'    => $this->request->getPost('clientobservation'),
                'technician'            => $this->request->getPost('technician'),
                'client_name'           => $this->request->getPost('clientname'),
                'status'                => $this->request->getPost('status')
            ]);

            if ($serviceorder->save()) {
                return $this->response->redirect('serviceorders');
            }

            $this->alert->error([
                'Erro!',
                $serviceorder->getMessages()[0]
            ]);
        }
    }`

I run using Xdebug and it receives the correct value from $_POST, it assigns too, but after it saves the field service_order_type_id reverts to the original value.

If I only change the field service_order_type_id the record doesn´t update. The field updated_at doesn´t change the value and it givves me no errors.

I know that the model is right because in the create method it´s working fine. In the debug it receives and assign a new value, but don´t know what´s happening when I try to save. The other fields updates without any errors.

**Comment by borisdelev** [comment]
Did you try to debug all post data? I see in post you get $this->request->getPost('serviceordertype')... For id field. Are you sure its correct?

Just for test- try simple update with one field and just change that field to existing type id.

**Reply by fctoscano** [reply]
Here´s some images

![debug_1](https://github.com/phalcon/phalcon/assets/10601958/cd17ed2e-7eda-43f6-a437-4a0c410016ab)
![debug_2](https://github.com/phalcon/phalcon/assets/10601958/4351217b-059d-4385-8a88-c810cd305443)
![debug_3](https://github.com/phalcon/phalcon/assets/10601958/9f88b222-a1fc-4279-811a-7cf5b22cd146)
![debug_post](https://github.com/phalcon/phalcon/assets/10601958/ac4efb04-4736-4934-bcf6-9c11a71ce48c)


I also tried to change the the line below:

`'service_order_type_id' => 6,`

But I keep getting the same results. If I cahnge only the service_order_type_id the record does not update, If i change the other fields the record is updated but the service_order_type_id keeps its value.

**Reply by fctoscano** [reply]
Running windows with apache 2.4.55, phalcon 5.2.1, php 8.1.9

**Reply by borisdelev** [reply]
Strange. All seems alright. Can you check relation... mean "service_order_type_id" is related with table "service_order_type_id". Does 6 exists there?
If... this is fine, can you provide me parts of your code-  export from your db just for that table (and relation) and the models. I will try on my local machine. But... all seems alright. Anyway :) Will find the issue. I think is something so small :)

**Reply by fctoscano** [reply]
Yes, the record exists.
I tried to simplify the code and the database but the error didn't appear. I'll try again and if I manage to duplicate the error I'll let you know.
If you want, we can do a remote login and you check the error directly in my computer.

**Reply by borisdelev** [reply]
Hmm.... we can try that too. I hope i can help.

Just something for the errors... because it can be with multiple items, you can implode it and see all... its just a hint:
```php
        $this->alert->error([
            'Erro!',
            implode(', ', $serviceorder->getMessages())
        ]);
```

**Reply by borisdelev** [reply]
Well... i try... and found interesting behaviour.... It appears when you invoke relation before save. Well... for example **save** method do save on ```$serviceorder``` and ```$serviceordertype``` at the same time. In your case:

```php
$type = $serviceorder->serviceordertype;
...
$serviceorder->save();
````

Still no idea why is that... ive read: https://docs.phalcon.io/5.0/en/db-models-relationships#save ...

**Comment by katesmith1304** [comment]
it will point to the new IP on the authoritative server[,](https://vidmateapp.win/) and recursive systems will update after their Cache based on the record TTL has expired[.](https://momix.uno/)[.](https://pikashow.bio/)..
